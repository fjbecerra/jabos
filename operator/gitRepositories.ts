import { Request, Response, NextFunction } from 'express';
import gitRepositorySshSecretEnv from './gitRepositorySshSecretEnv';
import jabosOperatorUrlEnv from './jabosOperatorUrlEnv';
import settings from './settings';

export default {
  async sync(request: Request, response: Response, next: NextFunction) {
    if (settings.debug()) console.log("gitRepositories sync req", JSON.stringify(request.body));

    var name: string = request.body.object.metadata.name;
    var namespace: string = request.body.object.metadata.namespace;
    var repo: any = request.body.object.spec;

    var jobName = `git-repository-updater-${name}-${Math.floor(Math.random() * 10e6)}i`;

    var jobs = Object.keys(request.body.attachments['Job.batch/v1']);
    if (jobs.length > 0) {
      var job = request.body.attachments['Job.batch/v1'][jobs[0]];
      if (job.status.succeeded !== 1) {
        jobName = job.metadata.name;
      }
    }

    var res = {
      "attachments": [
        {
          "apiVersion": "rbac.authorization.k8s.io/v1",
          "kind": "Role",
          "metadata": {
            "name": `builder-${name}`,
            "namespace": namespace,
          },
          "rules": [
            {
              "apiGroups": ["jabos.io"],
              "resources": ["docker-images", "jsonnet-manifests", "git-repositories"],
              "verbs": ["get", "list", "watch", "patch"],
            }
          ],
        },
        {
          "apiVersion": "rbac.authorization.k8s.io/v1",
          "kind": "RoleBinding",
          "metadata": {
            "name": `builder-${name}`,
            "namespace": namespace,
          },
          "roleRef": {
            "apiGroup": "rbac.authorization.k8s.io",
            "kind": "Role",
            "name": `builder-${name}`,
          },
          "subjects": [
            {
              "kind": "ServiceAccount",
              "namespace": namespace,
              "name": `builder-${name}`,
            }
          ],
        },
        {
          "apiVersion": 'v1',
          "kind": 'ServiceAccount',
          "metadata": {
            "name": `builder-${name}`,
            "namespace": namespace,
          },
        },
        {
          "apiVersion": "batch/v1",
          "kind": "Job",
          "metadata": {
            "name": jobName,
            "labels": {
              "type": "git-repository-updater"
            },
          },
          "spec": {
            "completions": 1,
            "completionMode": "NonIndexed",
            "backoffLimit": 100,
            "activeDeadlineSeconds": parseInt(settings.jobActiveDeadlineSeconds()),
            "ttlSecondsAfterFinished": 30,
            "parallelism": 1,
            "template": {
              "metadata": {
                "name": jobName,
                "labels": {
                  "job": jobName
                }
              },
              "spec": {
                "serviceAccountName": `builder-${name}`,
                "restartPolicy": "OnFailure",
                "securityContext": {
                  "runAsNonRoot": true,
                },
                "containers": [
                  {
                    "image": `${settings.imagePrefix()}git-repository-updater:${settings.buildNumber()}`,
                    "args": [repo.url, repo.branch, namespace, name],
                    "env": [...jabosOperatorUrlEnv(), ...gitRepositorySshSecretEnv(repo.ssh)],
                    "name": "git-repository-updater",
                    "imagePullPolicy": settings.imagePullPolicy(),
                    "securityContext": {
                      "readOnlyRootFilesystem": true,
                      "allowPrivilegeEscalation": false,
                      "runAsNonRoot": true,
                      "capabilities": {
                        "drop": ['ALL'],
                      },
                    },
                    "resources": {
                      "limits": {
                        "cpu": "500m",
                        "memory": "500Mi"
                      },
                      "requests": {
                        "cpu": "100m",
                        "memory": "100Mi"
                      }
                    },
                    "volumeMounts": [
                      {
                        "name": "git-temp",
                        "mountPath": "/gitTemp",
                      },
                      {
                        "name": "temp",
                        "mountPath": "/tmp",
                      }
                    ]
                  }
                ],
                volumes: [
                  {
                    "name": "git-temp",
                    "emptyDir": {}
                  },
                  {
                    "name": "temp",
                    "emptyDir": {}
                  },
                ]
              }
            }
          }
        }
      ],
    };

    if (settings.debug()) console.log("gitRepositories sync res", JSON.stringify(res));
    response.status(200).json(res);
  },

  async customize(request: Request, response: Response, next: NextFunction) {
    if (settings.debug()) console.log("gitRepositories customize req", JSON.stringify(request.body));

    var res = {
      "relatedResources": []
    };

    if (settings.debug()) console.log("gitRepositories customize res", JSON.stringify(res));
    response.status(200).json(res);
  },

  async finalize(request: Request, response: Response, next: NextFunction) {
    if (settings.debug()) console.log("gitRepositories finalize req", JSON.stringify(request.body));

    var res = {
      "annotations": {},
      "attachments": [],
      "finalized": true,
    }

    if (settings.debug()) console.log("gitRepositories finalize res", JSON.stringify(res));
    response.status(200).json(res);
  }
}
