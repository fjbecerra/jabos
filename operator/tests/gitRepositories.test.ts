import { Request, Response } from 'express';
import mockResponse from './mockResponse';
import gitRepositories from '../gitRepositories';
import { setMock, clearMock } from './settingsMock';

beforeEach(() => {
  setMock();
});

afterEach(() => {
  clearMock();
});

test('gitRepositories sync', async () => {
  const req = {
    body: {
      object: {
        metadata: {
          name: 'name_value',
          uid: 'uid_value',
          namespace: 'namespace_value'
        },
        spec: {
          url: 'url_value',
          branch: 'branch_value'
        },
        status: {
          lastCommit: 'lastCommit_value'
        }
      },
    }
  } as unknown as Request;
  const res = mockResponse();
  await gitRepositories.sync(req, res as unknown as Response, () => { /* NOOP */ });
  expect(res.json.mock.calls[0]).toMatchSnapshot();
  expect(res.status).toHaveBeenCalledWith(200);
});

test('gitRepositories sync with promotedCommit', async () => {
  const req = {
    body: {
      object: {
        metadata: {
          name: 'name_value',
          namespace: 'namespace_value'
        },
        spec: {
          promotedCommit: 'promotedCommit_value'
        }
      },
    }
  } as unknown as Request;
  const res = mockResponse();
  await gitRepositories.sync(req, res as unknown as Response, () => { /* NOOP */ });
  expect(res.json.mock.calls[0]).toMatchSnapshot();
  expect(res.status).toHaveBeenCalledWith(200);
});

test('gitRepositories customize', async () => {
  const req = {} as unknown as Request;
  const res = mockResponse();
  await gitRepositories.customize(req, res as unknown as Response, () => { /* NOOP */ });
  expect(res.json.mock.calls[0]).toMatchSnapshot();
  expect(res.status).toHaveBeenCalledWith(200);
});

test('gitRepositories finalize', async () => {
  const req = {} as unknown as Request;
  const res = mockResponse();
  await gitRepositories.finalize(req, res as unknown as Response, () => { /* NOOP */ });
  expect(res.json.mock.calls[0]).toMatchSnapshot();
  expect(res.status).toHaveBeenCalledWith(200);
});
