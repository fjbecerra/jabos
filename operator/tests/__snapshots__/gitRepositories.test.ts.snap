// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gitRepositories customize 1`] = `
Array [
  Object {
    "relatedResources": Array [],
  },
]
`;

exports[`gitRepositories finalize 1`] = `
Array [
  Object {
    "annotations": Object {},
    "attachments": Array [],
    "finalized": true,
  },
]
`;

exports[`gitRepositories sync 1`] = `
Array [
  Object {
    "attachments": Array [
      Object {
        "apiVersion": "rbac.authorization.k8s.io/v1",
        "kind": "Role",
        "metadata": Object {
          "name": "builder-name_value",
          "namespace": "namespace_value",
        },
        "rules": Array [
          Object {
            "apiGroups": Array [
              "jabos.io",
            ],
            "resources": Array [
              "docker-images/status",
              "jsonnet-manifests/status",
              "git-repositories/status",
            ],
            "verbs": Array [
              "patch",
            ],
          },
          Object {
            "apiGroups": Array [
              "events.k8s.io",
            ],
            "resources": Array [
              "events",
            ],
            "verbs": Array [
              "get",
              "list",
              "watch",
              "create",
              "update",
              "patch",
            ],
          },
        ],
      },
      Object {
        "apiVersion": "v1",
        "kind": "ServiceAccount",
        "metadata": Object {
          "name": "builder-name_value",
          "namespace": "namespace_value",
        },
      },
      Object {
        "apiVersion": "rbac.authorization.k8s.io/v1",
        "kind": "RoleBinding",
        "metadata": Object {
          "name": "builder-name_value",
          "namespace": "namespace_value",
        },
        "roleRef": Object {
          "apiGroup": "rbac.authorization.k8s.io",
          "kind": "Role",
          "name": "builder-name_value",
        },
        "subjects": Array [
          Object {
            "kind": "ServiceAccount",
            "name": "builder-name_value",
            "namespace": "namespace_value",
          },
        ],
      },
      Object {
        "apiVersion": "batch/v1",
        "kind": "Job",
        "metadata": Object {
          "labels": Object {
            "type": "git-repository-updater",
          },
          "name": "git-repository-updater-name_value",
        },
        "spec": Object {
          "activeDeadlineSeconds": 1234,
          "backoffLimit": 1234,
          "completionMode": "NonIndexed",
          "completions": 1,
          "parallelism": 1,
          "template": Object {
            "metadata": Object {
              "labels": Object {
                "job": "git-repository-updater-name_value",
              },
              "name": "git-repository-updater-name_value",
            },
            "spec": Object {
              "containers": Array [
                Object {
                  "args": Array [],
                  "env": Array [
                    Object {
                      "name": "URL",
                      "value": "url_value",
                    },
                    Object {
                      "name": "BRANCH",
                      "value": "branch_value",
                    },
                    Object {
                      "name": "NAMESPACE",
                      "value": "namespace_value",
                    },
                    Object {
                      "name": "NAME",
                      "value": "name_value",
                    },
                    Object {
                      "name": "OBJECT_UID",
                      "value": "uid_value",
                    },
                    Object {
                      "name": "CURRENT_COMMIT",
                      "value": "lastCommit_value",
                    },
                    Object {
                      "name": "JABOS_OPERATOR_URL",
                      "value": "http://operator.settings_namespace:1234/",
                    },
                  ],
                  "image": "settings_imagePrefixgit-repository-updater:settings_build_number",
                  "imagePullPolicy": "Always",
                  "name": "git-repository-updater",
                  "resources": Object {
                    "limits": Object {
                      "cpu": "500m",
                      "memory": "500Mi",
                    },
                    "requests": Object {
                      "cpu": "100m",
                      "memory": "100Mi",
                    },
                  },
                  "securityContext": Object {
                    "allowPrivilegeEscalation": false,
                    "capabilities": Object {
                      "drop": Array [
                        "ALL",
                      ],
                    },
                    "readOnlyRootFilesystem": true,
                    "runAsNonRoot": true,
                  },
                  "stdin": true,
                  "tty": true,
                  "volumeMounts": Array [
                    Object {
                      "mountPath": "/gitTemp",
                      "name": "git-temp",
                    },
                    Object {
                      "mountPath": "/tmp",
                      "name": "temp",
                    },
                  ],
                },
              ],
              "restartPolicy": "Never",
              "securityContext": Object {
                "runAsNonRoot": true,
              },
              "serviceAccountName": "builder-name_value",
              "volumes": Array [
                Object {
                  "emptyDir": Object {},
                  "name": "git-temp",
                },
                Object {
                  "emptyDir": Object {},
                  "name": "temp",
                },
              ],
            },
          },
          "ttlSecondsAfterFinished": 1234,
        },
      },
    ],
  },
]
`;

exports[`gitRepositories sync with promotedCommit 1`] = `
Array [
  Object {
    "attachments": Array [
      Object {
        "apiVersion": "rbac.authorization.k8s.io/v1",
        "kind": "Role",
        "metadata": Object {
          "name": "builder-name_value",
          "namespace": "namespace_value",
        },
        "rules": Array [
          Object {
            "apiGroups": Array [
              "jabos.io",
            ],
            "resources": Array [
              "docker-images/status",
              "jsonnet-manifests/status",
              "git-repositories/status",
            ],
            "verbs": Array [
              "patch",
            ],
          },
          Object {
            "apiGroups": Array [
              "events.k8s.io",
            ],
            "resources": Array [
              "events",
            ],
            "verbs": Array [
              "get",
              "list",
              "watch",
              "create",
              "update",
              "patch",
            ],
          },
        ],
      },
      Object {
        "apiVersion": "v1",
        "kind": "ServiceAccount",
        "metadata": Object {
          "name": "builder-name_value",
          "namespace": "namespace_value",
        },
      },
      Object {
        "apiVersion": "rbac.authorization.k8s.io/v1",
        "kind": "RoleBinding",
        "metadata": Object {
          "name": "builder-name_value",
          "namespace": "namespace_value",
        },
        "roleRef": Object {
          "apiGroup": "rbac.authorization.k8s.io",
          "kind": "Role",
          "name": "builder-name_value",
        },
        "subjects": Array [
          Object {
            "kind": "ServiceAccount",
            "name": "builder-name_value",
            "namespace": "namespace_value",
          },
        ],
      },
    ],
    "status": Object {
      "conditions": Array [
        Object {
          "status": "True",
          "type": "Syncing",
        },
      ],
      "latestCommit": "promotedCommit_value",
    },
  },
]
`;
